lib/spack/spack/traverse.py

```diff
@@ -0,0 +1,328 @@
+            for example BaseVisitor.
+    """
+    queue = init_queue(specs)
+    while len(queue) > 0:
+        node = queue.pop(0)
+
+        # If the visitor doesn't accept the node, we don't traverse it further.
+        if not visitor.accept(node):
+            continue
+
+        for edge in visitor.neighbors(node):
+            queue.append(EdgeAndDepth(edge, node.depth + 1))
+
+
+# Breadth first traversal to trees
+
+
+def breadth_first_to_tree_edges(roots, deptype="all"):
+    """This produces an adjacency list (with edges) and a map of parents.
+    There may be nodes that are reached through multiple edges. To print as
+    a tree, one should use the parents dict to verify if the path leading to
+    the node is through the correct parent. If not, the branch should be
+    truncated."""
+    edges = defaultdict(list)
+    parents = dict()
+
+    for edge in traverse_breadth_first_edges(roots, cover="edges", deptype=deptype):
+        parent_id = None if edge.parent is None else edge.parent.dag_hash()
+        child_id = edge.spec.dag_hash()
+        edges[parent_id].append(edge)
+        if child_id not in parents:
+            parents[child_id] = parent_id
+
+    return edges, parents
+
+
+def breadth_first_to_tree_nodes(roots, deptype="all"):
+    """This produces a list of edges that forms a tree; every node has no more
+    that one incoming edge."""
+    edges = defaultdict(list)
+
+    for edge in traverse_breadth_first_edges(roots, cover="nodes", deptype=deptype):
+        parent_id = None if edge.parent is None else edge.parent.dag_hash()
+        edges[parent_id].append(edge)
+
+    return edges
+
+
+def traverse_breadth_first_tree_edges(parent_id, edges, parents, depth=0):
+    """Do a depth-first search on edges generated by bread-first traversal,
+    which can be used to produce a tree."""
+    for edge in edges[parent_id]:
+        yield (depth, edge)
+
+        child_id = edge.spec.dag_hash()
+
+        # Don't follow further if we're not the parent
+        if parents[child_id] != parent_id:
+            continue
+
+        # yield from ... in Python 3.
+        for item in traverse_breadth_first_tree_edges(child_id, edges, parents, depth + 1):
+            yield item
+
+
+def traverse_breadth_first_tree_nodes(parent_id, edges, depth=0):
+    for edge in edges[parent_id]:
+        yield (depth, edge)
+        for item in traverse_breadth_first_tree_nodes(edge.spec.dag_hash(), edges, depth + 1):
+            yield item
+
+
+def traverse_breadth_first_tree(specs, cover="nodes", deptype="all"):
+    if cover == "edges":
+        edges, parents = breadth_first_to_tree_edges(specs, deptype)
+        generator = traverse_breadth_first_tree_edges(None, edges, parents)
+    elif cover == "nodes":
+        edges = breadth_first_to_tree_nodes(specs, deptype)
+        generator = traverse_breadth_first_tree_nodes(None, edges)
+    else:
+        raise ValueError("cover should be nodes or edges")
```

Comment:
It can make sense to use `traverse_breadth_first_edges` to iterate over all edges in BFS order without pruning; it doesn't make sense to print a tree this way, since it's the same order that DFS would give, so it would be the same as DFS except that it's doing a redundant traversal in BFS order first.